deepchem/models/xgboost_models/__init__.py

```diff
@@ -0,0 +1,121 @@
+"""
+Scikit-learn wrapper interface of xgboost
+"""
+
+import xgboost as xgb
+import numpy as np
+import os
+from deepchem.models import Model
+from deepchem.models.sklearn_models import SklearnModel
+from deepchem.utils.save import load_from_disk
+from deepchem.utils.save import save_to_disk
+from sklearn.cross_validation import train_test_split
+from sklearn.grid_search import GridSearchCV
+import tempfile
+
+
+class XGBoostModel(SklearnModel):
+  """
+  Abstract base class for XGBoost model.
+  """
+
+  def __init__(self,
+               model_instance=None,
+               model_dir=None,
+               verbose=False,
+               **kwargs):
+    """Abstract class for XGBoost models.
+    Parameters:
+    -----------
+    model_instance: object
+      Scikit-learn wrapper interface of xgboost
+    model_dir: str
+      Path to directory where model will be stored.
+    """
+    if model_dir is not None:
+      if not os.path.exists(model_dir):
+        os.makedirs(model_dir)
+    else:
+      model_dir = tempfile.mkdtemp()
+    self.model_dir = model_dir
+    self.model_instance = model_instance
+    self.model_class = model_instance.__class__
+
+    self.verbose = verbose
+    if 'early_stopping_rounds' in kwargs:
+      self.early_stopping_rounds = kwargs['early_stopping_rounds']
+    else:
+      self.early_stopping_rounds = 50
+
+  def fit(self, dataset, **kwargs):
+    """
+    Fits XGBoost model to data.
+    """
+    X = dataset.X
+    y = np.squeeze(dataset.y)
+    w = np.squeeze(dataset.w)
+    seed = self.model_instance.seed
+    if isinstance(self.model_instance, xgb.XGBClassifier):
+      xgb_metric = "auc"
+      sklearn_metric = "roc_auc"
+      stratify = y
+    elif isinstance(self.model_instance, xgb.XGBRegressor):
+      xgb_metric = "mae"
+      sklearn_metric = "neg_mean_absolute_error"
+      stratify = None
+    best_param = self._search_param(sklearn_metric, X, y)
+    # update model with best param
+    self.model_instance = self.model_class(**best_param)
+
+    # Find optimal n_estimators based on original learning_rate
+    # and early_stopping_rounds
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.2, random_state=seed, stratify=stratify)
+
+    self.model_instance.fit(
+        X_train,
+        y_train,
+        early_stopping_rounds=self.early_stopping_rounds,
+        eval_metric=xgb_metric,
+        eval_set=[(X_train, y_train), (X_test, y_test)],
+        verbose=self.verbose)
+    # Since test size is 20%, when retrain model to whole data, expect
+    # n_estimator increased to 1/0.8 = 1.25 time.
+    estimated_best_round = np.round(self.model_instance.best_ntree_limit * 1.25)
+    self.model_instance.n_estimators = np.int64(estimated_best_round)
+    self.model_instance.fit(
```

Comment:
Seems that you are directly doing hyperparameter search when executing the model. Will this influence the overall speed? Also, as we do not implement detailed hyperparameter optimization(same settings for all dataset) for other deepchem models, this could be an issue when comparing between models. @rbharath What's your idea about this?